File: Server/Spring_Server/src/main/java/com/Spacegram/server/utility/Hash/RSAHash.java
package com.Spacegram.server.utility.Hash;


import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import javax.crypto.Cipher;

public class RSAHash {
    private KeyPairGenerator keyPairGenerator;
    private KeyPair keyPair;
    private PublicKey publicKey;
    private PrivateKey privateKey;

    public RSAHash() throws Exception {
        keyPairGenerator = KeyPairGenerator.getInstance("RSA");
        keyPairGenerator.initialize(2048);
        keyPair = keyPairGenerator.generateKeyPair();
        publicKey = keyPair.getPublic();
        privateKey = keyPair.getPrivate();
    }

    public String generatePublicKeys() {
        return Base64.getEncoder().encodeToString(publicKey.getEncoded());
    }

    public String generatePrivateKeys() {
        return Base64.getEncoder().encodeToString(privateKey.getEncoded());
    }

    public String encrypt(String key, String data) throws Exception {
        byte[] publicKeyBytes = Base64.getDecoder().decode(key);
        X509EncodedKeySpec keySpec = new X509EncodedKeySpec(publicKeyBytes);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        PublicKey publicKey = keyFactory.generatePublic(keySpec);

        Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
        cipher.init(Cipher.ENCRYPT_MODE, publicKey);
        byte[] encryptedBytes = cipher.doFinal(data.getBytes("UTF-8"));
        return Base64.getEncoder().encodeToString(encryptedBytes);
    }

    public String decrypt(byte[] data, String Key) throws Exception {
        byte[] privateKeyBytes = Base64.getDecoder().decode(Key);
        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(privateKeyBytes);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        PrivateKey privateKey = keyFactory.generatePrivate(keySpec);

        Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
        cipher.init(Cipher.DECRYPT_MODE, privateKey);
        byte[] decryptedData = cipher.doFinal(data);
        return new String(decryptedData, "UTF-8");
    }
}

File: Server/Csharp_Server/UserServer/Hash/RSAHash.cs
using UserServer.Interface.Hash;
using System;
using System.Security.Cryptography;
using System.Text;

namespace UserServer.Hash
{
    internal class RSAHash : IRSAHash
    {
        private RSA _rsa;

        public RSAHash()
        {
            _rsa = RSA.Create();
        }

        public string GeneratePublicKeys()
        {
            var publicKey = Convert.ToBase64String(_rsa.ExportSubjectPublicKeyInfo());
            return (publicKey);
        }
        
        public string GeneratePrivateKeys()
        {
            var privateKey = Convert.ToBase64String(_rsa.ExportPkcs8PrivateKey());
            return ( privateKey);
        }

        public byte[] Encrypt(string data, string publicKey)
        {
            _rsa.ImportSubjectPublicKeyInfo(Convert.FromBase64String(publicKey), out _);
            return _rsa.Encrypt(Encoding.UTF8.GetBytes(data), RSAEncryptionPadding.OaepSHA256);
        }


        public string Decrypt(byte[] data, string privateKey)
        {
            _rsa.ImportPkcs8PrivateKey(Convert.FromBase64String(privateKey), out _);
            var decryptedData = _rsa.Decrypt(data, RSAEncryptionPadding.OaepSHA256);
            return Encoding.UTF8.GetString(decryptedData);
        }
    }
}
